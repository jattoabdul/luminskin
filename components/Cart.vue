<template>
  <div>
    <h3>Shopping Cart</h3>
    <div class="cart">
      <div class="panel panel-default">
        <div class="panel-header">
          <!-- currency selector -->
          <p @click.stop="toggleCurrencySelector()">
            Select Currency
          </p>
          <ul v-show="showCurrencySelector" class="currency-selector">
            <li v-for="(eachCurrency, index) in currency" :key="index" @click.stop="selectCurrency(index)">
              {{ eachCurrency }} - currency
            </li>
          </ul>
          <!-- end of currency selector -->
        </div>
        <div class="panel-body">
          <div v-if="computedCart.length > 0" class="cart__body">
            <ul>
              <CartItem
                v-for="item in computedCart"
                :key="item.id"
                :item="item"
                :current-currency="currentCurrency"
              />
            </ul>
          </div>
          <div v-if="computedCart.length <= 0" class="alert alert-info">
            Cart is empty
          </div>
          <div class="cart__total">
            Subtotal: {{ cartSubTotal | currency({symbol: getCurrencySymbol(currentCurrency.toLowerCase())}) }}
          </div>
          <button @click.stop="clearCart()">
            Clear Cart
          </button>
          <button @click.stop="proceedToCheckout()" :disabled="computedCart.length <= 0">
            Proceed To checkout
          </button>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import CartItem from './CartItem.vue'

export default {
  name: 'Cart',
  components: {
    CartItem
  },
  props: {
    showCurrencySelector: {
      type: Boolean,
      default() {
        return false
      }
    },
    currency: {
      type: Array,
      default() {
        return []
      }
    },
    cart: {
      type: Array,
      default() {
        return []
      }
    },
    currentCurrency: {
      type: String,
      default() {
        return []
      }
    },
    tax: {
      type: Number,
      default() {
        return 0.065
      }
    },
    products: {
      type: Array,
      default() {
        return []
      }
    }
  },
  computed: {
    cartSubTotal() { return this.computedCart.reduce((acc, cur) => acc + (cur.price * cur.quantity), 0) },
    cartTotal() { return this.cartSubTotal + (this.tax * this.cartSubTotal) },
    computedCart() {
      return this.cart.map((cartproduct) => {
        const foundProduct = this.products.find((itemForSale) => {
          return cartproduct.id === itemForSale.id
        })
        foundProduct.quantity = cartproduct.quantity
        return foundProduct
      })
    }
  },
  methods: {
    selectCurrency(index) {
      this.$eventBus.$emit('select-currency', { index })
    },
    clearCart() {
      this.$eventBus.$emit('clear-cart')
    },
    proceedToCheckout() {
      this.$eventBus.$emit('proceed-to-checkout')
    },
    toggleCurrencySelector() {
      this.$eventBus.$emit('toggle-currency-selector')
    }
  }
}
</script>

<style scoped lang="scss">
.cart__body {
  margin-bottom: 0.7em;
}
.cart__total {
  font-weight: bold;
  font-size: 1.5em;
  line-height: 1.1em;
  text-align: right;
}
</style>
