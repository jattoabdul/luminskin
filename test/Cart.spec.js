import { shallowMount } from '@vue/test-utils'
import Cart from '@/components/Cart.vue'
import Vue from 'vue'
import VueCurrencyFilter from 'vue-currency-filter'
import getSymbolFromCurrency from 'currency-symbol-map'

Vue.use(VueCurrencyFilter, {
  thousandsSeparator: ',',
  fractionCount: 2,
  fractionSeparator: '.',
  symbolPosition: 'front',
  symbolSpacing: false
})

Vue.mixin({
  methods: {
    getCurrencySymbol(currency) {
      const foundCurrency = getSymbolFromCurrency(currency)
      if (foundCurrency) {
        return foundCurrency
      }
      return currency
    }
  }
})

// @TODO: Fix Tests
const products = [
  {
    id: 1,
    title: 'product title',
    price: 100,
    image_url: 'image.svg'
  }
]
const currency = ['USD', 'NGN']

const factory = () => {
  return shallowMount(Cart, {
    propsData: {
      showCurrencySelector: false,
      currency,
      cart: [],
      currentCurrency: 'USD',
      tax: 0.065,
      showCart: false,
      products
      // cartSubTotal: 0,
      // cartTotal: [],
      // computedCart: []
    }
  })
}

describe('Cart', () => {
  test('mounts properly', () => {
    const wrapper = factory()
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders correctly', () => {
    // const wrapper = mount(Cart)
    const wrapper = factory()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
