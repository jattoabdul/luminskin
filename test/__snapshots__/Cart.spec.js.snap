// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders correctly 1`] = `
<div class="cart__widget">
  <div class="cart__content">
    <header><span class="go-back"><img src="/back.svg" alt="go-back-to-products"></span>
      <h3 class="cart__heading">
        Your Cart
      </h3>
      <div class="currency-selection"><select name="currency">
          <option value="0">
            USD
          </option>
          <option value="1">
            NGN
          </option>
        </select></div>
    </header>
    <div class="cart__items-wrapper">
      <!---->
      <div class="alert alert-info">
        Cart is empty
      </div>
    </div>
    <div class="cart__footer">
      <div class="cart__total"><span>Subtotal</span> <span> $0.00</span></div> <button class="btn">
        Make this a subscription (save 20%)
      </button> <button disabled="disabled" class="btn btn-primary">
        Proceed To checkout
      </button>
    </div>
  </div>
</div>
`;
